### PR TITLE
181 add map validation

### LINKED ISSUE #181: add map validation
- create function isMapValid() in Floorplan
-  check all characters are valid. (#, ., S)
- BFS from starting point  till obstacle cells met (if valid) or out of boundary (if invalid)

### PR DESCRIPTION
**Pull request by Raa37**

### CODE REVIEW

File: Floorplan.py [hunk 1]
```diff
@@ -8,7 +8,8 @@ class Floorplan(object):
     '''
     
     def __init__(self, drawing):
-        self.width, self.height, self.obstacles = self._parseDrawing(drawing)
+        assert self._isMapValid(drawing)
+        self.width, self.height, self.obstacles, self.initX, self.initY = self._parseDrawing(drawing)
```
Comment (twatteyne) [inline Floorplan.py]: add explicit tuple: `(self.width, self.height, self.obstacles, self.initX, self.initY)`

File: Floorplan.py [hunk 2]
```diff
@@ -37,6 +41,70 @@ def _parseDrawing(drawing):
...
+                    assert matrixOfChars[nx][ny] == '#' or matrixOfChars[nx][ny] == '.' or matrixOfChars[nx][ny] == 's'
+
+                    # if next frontier position is not in matrix of characters map is invalid
+                    if matrixOfChars[nx][ny] != '#' and (nx, ny) not in charsToCheck  and (nx, ny) not in charsChecked:
+                        charsToCheck += [(nx, ny)]
+
+                except IndexError:
+                    returnVal = False
 
+        return returnVal
```
Comment (twatteyne) [inline Floorplan.py]: add empty newline at end of file

File: Floorplan.py [hunk 3]
```diff
@@ -37,6 +41,70 @@ def _parseDrawing(drawing):
...
+                    \'''
+            },
+            'out': False
+        },
+        '''
+
+        returnVal       = True
+
+        # split drawing into matrix of characters indexed by row, column indices
+        matrixOfChars   = [line for line in drawing.splitlines() if line]
```
Comment (twatteyne) [inline Floorplan.py]: why "if line"?

File: Floorplan.py [hunk 4]
```diff
@@ -37,6 +41,70 @@ def _parseDrawing(drawing):
...
+                (charRow,    charCol-1), (charRow,    charCol+1),
+                (charRow-1,  charCol-1), (charRow+1,  charCol+1),
+                (charRow-1,  charCol+1), (charRow+1,  charCol-1)
+            ]
+
+            for nx, ny in neighbourChars:
+                
+                try:
+                    # if drawing contains characters that aren't '#,.,s' map is invalid
+                    assert matrixOfChars[nx][ny] == '#' or matrixOfChars[nx][ny] == '.' or matrixOfChars[nx][ny] == 's'
```
Comment (twatteyne) [inline Floorplan.py]: do we draw the map with dots of spaces?

File: Floorplan.py [hunk 5]
```diff
@@ -37,6 +41,70 @@ def _parseDrawing(drawing):
...
+        charsToCheck    = []     # character positions in map of which surrounding frontiers are to be determined
+
+        # if there is no starting position character 's' map is invalid
+        try:
+            startPos    = [(lidx,line.index('s')) for (lidx,line) in enumerate(matrixOfChars) if 's' in line][0]
+        except IndexError:
+            startPos    = None
+            returnVal   = False
+
+        charsToCheck   += [startPos]
```
Comment (twatteyne) [inline Floorplan.py]: replace rename chars by cells

**Review by twatteyne — CHANGES_REQUESTED** [review]
excellent work on this one. Some super small comments.

File: Floorplan.py (hunk 4 above)
Comment (Raa37) [reply Floorplan.py]: I thought it's best to use dots as it's easier to differentiate between empty spaces (that may be there by mistake, wrong indentation, .. etc) and the actual free space in our maps. it also helps with debugging

**Review by twatteyne — APPROVED** [review]

**Pull request merged**